nova/virt/vmwareapi/special_spawning.py

```diff
@@ -0,0 +1,218 @@
+        # no group -> nothing to do
+        if group is None:
+            return True
+
+        # if there are no hosts in the group, we're done
+        if not getattr(group, 'host', None):
+            return True
+
+        group.host = []
+
+        client_factory = self._session.vim.client.factory
+        group_spec = client_factory.create('ns0.ClusterGroupSpec')
+        group_spec.operation = 'edit'
+        group_spec.info = group
+
+        config_spec = client_factory.create('ns0:ClusterConfigSpecEx')
+        config_spec.groupSpec = [group_spec]
+        cluster_util.reconfigure_cluster(self._session, self._cluster,
+                                         config_spec)
+
+        return True
+
+    def free_host(self, context):
+        """Find a host that doesn't have a bigvm and put it into the special
+        hostgroup. If that's already the case, return whether there are running
+        VMs left on the host, i.e. the process is finished.
+        """
+        # get the group
+        group = self._get_group()
+
+        if group is None or not getattr(group, 'host', None):
+            # find a host to free
+            # get all the vms in a cluster, because we need to find a host
+            # without big VMs. Take the one with least memory used.
+            props = ['config.hardware.memoryMB', 'runtime.host',
+                     'runtime.powerState']
+            cluster_vms = self._vmops._list_instances_in_cluster(props)
+            if not cluster_vms:
+                LOG.error('Got no VMs for freeing a host for spawning. '
+                          'Treating as error instead of continuing, as an '
+                          'empty cluster is unlikely.')
+                return FREE_HOST_STATE_ERROR
+
+            vms_per_host = {}
+            for vm_uuid, vm_props in cluster_vms:
+                props = (vm_props.get('config.hardware.memoryMB', 0),
+                         vm_props.get('runtime.powerState', 'poweredOff'))
+                vms_per_host.setdefault(vm_props.get('runtime.host'), []). \
+                        append(props)
+
+            # filter for hosts without big VMs
+            # FIXME ask david if we use big VMs here or the smaller special
+            # spawning vms
+            vms_per_host = {h: vms for h, vms in vms_per_host.items()
+                            if all(mem < CONF.bigvm_mb for mem, state in vms)}
+
+            if not vms_per_host:
+                LOG.warning('No suitable host found for freeing a host for '
+                            'spawning.')
+                return FREE_HOST_STATE_ERROR
+
+            mem_per_host = {h: sum(mem for mem, state in vms
+                                   if state != 'poweredOff')
+                            for h, vms in vms_per_host.items()}
+
+            # take the one with least memory used
+            host_ref, _ = sorted(mem_per_host.items(), key=lambda x, y: y,
+                                  reverse=True)[0]
```

Comment:
Afaik it's neither of both. It should be the configured memory of the VM. (We only reserve memory for big VMs.)

What are you aiming at with that question? Do you think we should use the current VM memory, because it would be easier to free a host with VMs not consuming much memory?

Reply:
I was trying to understand how it works, first of all) 
yes, if possible, I think that we should preferably decide here based on actual, current ram usage